metronome/hotstuff/service/src/io/iohk/metronome/hotstuff/service/messages/DuplexMessage.scala

```diff
@@ -0,0 +1,24 @@
+package io.iohk.metronome.hotstuff.service.messages
+
+import io.iohk.metronome.hotstuff
+import io.iohk.metronome.hotstuff.consensus.basic.Agreement
+
+/** Messages type to use in the networking layer if the use case has
+  * application specific message types, e.g. ledger synchronisation,
+  * not just the general BFT agreement (which could be enough if
+  * we need to execute all blocks to synchronise state).
+  */
+sealed trait DuplexMessage[A <: Agreement, M]
```

Comment:
Do we need the 2 type params?

Reply:
See my [other comment](https://github.com/input-output-hk/metronome/pull/23#discussion_r614724939) about what I think HotStuff setups would look like: they have the generic and the application specific part where we can dispatch messages to.

I think not all applications need their specific messages, though. For example I'd like to do an example application which just agrees on the order of some "commands", e.g. randomly generated strings. These wouldn't have a ledger for example, just a block history, like in the HotStuff paper, so they don't need the equivalent of the `CheckpointingMessage`. I could have moved this `DuplexMessage` into the `checkpointing` module, in which case it wouldn't need a generic parameter, however I thought that the `hotstuff` module should capture as much of the reusable patterns as possible, so the application specific part can be thin. 

This could be replaced with just an `Either`, but I thought it serves as better documentation of what a reusable pattern could be, in that you either instantiate a `RemoteConnectionManager[Task, PublicKey, HotStuffMessage[YourAgreement]]` or a `RemoteConnectionManager[Task, PublicKey, DuplexMessage[YourAgreement, YourMessages]]` and we can provide RLP formatters for the `DuplexMessage`, rather than all `Either` in general.

Reply:
What I don't get is how this union type is better than simple subclassing of a common `Message` class, especially that `M` is unbounded.

Reply:
As I understand there won't be a case where both `A` and `M` are defined as something other than `Nothing`, would they?

Reply:
What should subclass a common message class exactly, and where would the baseclass live?

Say I don't need application specific messages, I just want to use the `HotStuffMessage`. Should it be a sublcass of some higher level message type? What purpose would that serve in that scenario?

On the other hand, if I want to be delivered my application messages to my application service, again why should it inherit some class from HotStuff? It's the only kind that should arrive there. 

So, nothing should ever have to deal with the base class.

> there won't be a case where both A and M are defined as something other than Nothing, would they?

Sorry I don't understand, the case for checkpointing will be that both are not `Nothing`, `A` going to be the `CheckpointingAgreement` and `M` the `CheckpointingMessage`. I'm probably misunderstanding the question.

Reply:
Ok, let me put it another way. You mention that `DuplexMessage` could replaced by `Either`. So what additional decision is there to be made (and where) based on whether it's a `Left` or `Right`? Why can't that decision be made based on the message type alone?

Reply:
The decision that has to be made based on whether we're dealing with a `Left` or `Right` is whether an incoming network packet has to be delivered to the HotStuff components or the Checkpointing ones. 

There were other options: 
* First I thought I'll just add a generic `ApplicationMessage` to the `HotStuffMessage`, but the problem is that it's not necessary to have one.
* Or I thought the Checkpointing side could be in charge of routing to the HotStuff side, since it knows about them (the module dependencies allow that in that direction), but I wanted to keep the Checkpointing side thin, so if we had another variant there's nothing to copy paste.
* So I ended up with this: we have `Either[Consensus, Sync]` or we have `Either[[Either[Consensus, Sync], Application]`. The HotStuff module can support both with these types, allowing the application root to assemble the correct stack.

> Why can't that decision be made based on the message type alone?

We have to instantiate the network stack with some kind of message type that has a `Codec`. If I used a single `Message` type, perhaps an unsealed trait that is extended in the checkpointing module, I think I'd lose the help from the compiler to tell me if I forgot to implement something.

Reply:
Ok, I guess the need for the `Codec` justifies this design, but I still don't understand why the `DuplexMessage` has to carry the `Agreement` type. I doesn't look like `RemoteConnectionManager` knows anything about `Agreements`.

Also:
> the case for checkpointing will be that both are not Nothing, A going to be the CheckpointingAgreement and M the CheckpointingMessage

but the docstring says: "These will be wrapped in an `ApplicationMessage`" and `ApplicationMessage` is defined as `DuplexMessage[Nothing, M]`

Reply:
The `DuplexMessage` carries the agreement type because it's part of the HotStuff module and one of its cases is fixed to be a `HotStuffMessage`; it's sole purpose is to combine `HotStuffMessage` with some application specific message family, not to be a replacement of `Either` where both sides can change freely. The `RemoteConnectionManager` doesn't know anything about `Agreement` but it also doesn't know anything about `DuplexMessage`. I though `ExtendedHotStuffMessage` could be a name as well, or something similar, but landed on "duplex" as in a duplex house with two sides.

About `Nothing`: now I understand what you meant. I'll probably have to add a `+` before `M` and `A` at some point, but always use `Nothing` when that generic parameter doesn't appear in the case class itself, for example [here](https://github.com/input-output-hk/metronome/blob/develop/metronome/networking/src/io/iohk/metronome/networking/NetworkEvent.scala#L43-L48). Otherwise I'd always have to specify what the other type is when constructing, say, an `AgreementMessage`. This seems to be [common practice](https://github.com/fpinscala/fpinscala/blob/master/answers/src/main/scala/fpinscala/errorhandling/Either.scala#L27-L28). 

I thought you meant whether we'd create a `RemoteConnectionManager[PublicKey, DuplexMessage[SomeAgreement, Nothing]]`.

Reply:
Sure, I'm not questioning the practices. However, I'm still not clear on the type choice. Maybe because currently I can only see `DuplexMessage` as being passed to `RemoteConnectionManager` and it will become more clear when I see the messages being routed and interpreted. Sorry for being so inquisitive, I don't mean to hinder the progress here, just to build a solid understanding of the code.